### PR TITLE
[image_picker] fixes for iOS which doesn't present camera/albums with more complex navigation

### PR DESCRIPTION
**Pull request by chris-rutkowski**
## Description

Existing implementation is incorrect especially when adding project as a module with more complex presentation structure.
If window's root view controller has presented modally any other view controller, it cannot present any other any more. Only the top most controller can present other.

With this change, the image picker/camera controller is presented on the top most view controller.

## Related Issues

I found this issue raised by someone already:
https://github.com/flutter/flutter/issues/34787

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by chris-rutkowski** [comment]
@xster @mehmetf @cyanglaz @blasten I saw you merged some other PRs related to image picker recently, that's why I'm @-ing you. Is there anything more I can do here in this PR?

File: packages/image_picker/image_picker/pubspec.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: image_picker
 description: Flutter plugin for selecting images from the Android and iOS image
   library, and taking new pictures with the camera.
 homepage: https://github.com/flutter/plugins/tree/master/packages/image_picker/image_picker
-version: 0.6.6+3
+version: 0.6.7
```
Comment (cyanglaz) [inline packages/image_picker/image_picker/pubspec.yaml:5]: there is no public api change, we should only update the patch -> 0.6.6+4

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m [hunk 2]
```diff
@@ -25,33 +25,29 @@ @interface FLTImagePickerPlugin () <UINavigationControllerDelegate, UIImagePicke
...
-  }
-  return self;
-}
-
 - (UIImagePickerController *)getImagePickerController {
   return _imagePickerController;
 }
 
+- (UIViewController *)viewController {
+  UIViewController *topController = [UIApplication sharedApplication].keyWindow.rootViewController;
```
Comment (cyanglaz) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: this is great! By the way, while we are here, can we also make it work for navigation controller? (pushed viewcontrollers)

Comment (chris-rutkowski) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: this solution will work in this case as well.
According to the [documentation](https://developer.apple.com/documentation/uikit/uiimagepickercontroller) you shouldn't push UIImagePickerController, but present it modally instead.

So in the situation that FlutterViewController is pushed into an existing navigation controller, e.g.:

```
Window
  rootViewController
    someOtherViewController
      navigation controller (with some native view controller and FlutterVC)
         image picker controller 

not

Window
  rootViewController
    someOtherViewController
      navigation controller (with some native view controller and FlutterVC, image picker controller)
          
```

I was worried about iPad which for a long time required popover presentation for albums, but since iOS 13 it presents modally as well without an issue.

**Comment by chris-rutkowski** [comment]
@cyanglaz as requested changed version to 0.6.6+5

**Comment by chris-rutkowski** [comment]
@cyanglaz any more feedback? I've already updated changelog a few times due to conflicts - other PRs being merged.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m (hunk 2 above)
Comment (chris-rutkowski) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: @cyanglaz the solution I made will in all cases you are concerned about.

UINavigationController, UITabBarController they both manage internally some view controllers, **but** they themselves inherit from UIViewControllerr and either of them can be window's root view controller, or presented on some existing View Controller of any type.

The topViewController cannot be pushed in the navigation stack. In that case, UINavigationCoontroller is topViewController and some other push view controller is managed by UINavigationController. And with that arrangement, my solution will correctly display the Image Picker on top of the navigation controller.

**Trust me** - in my company we have already very complex navigation - soome legacy Drawer Controller (in iOS), multiple Navigation controllers, TabBarControllers and Flutter V.C. is displayed on top of all of this. And my solution has been proven successful.

**Comment by aboutlan** [comment]
@cyanglaz I have the same problem.
I presented flutterViewController after I use present image_picker doesn't work, Xcode error warning: Attempt to present <UIImagePickerController: 0x7fbbb1856800> on <UINavigationController: 0x7fbbb201c000> whose view is not in the window hierarchy !. 
chris-rutkowski solved my problem

**Comment by diegotori** [comment]
@cyanglaz a gentle reminder for reviewing this PR, since this could be very useful to have.

@chris-rutkowski looks like your branch has conflicts that need to be resolved before this can merge. Thank you in advance for this PR.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m (hunk 2 above)
Comment (cyanglaz) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: according to https://developer.apple.com/documentation/uikit/uiviewcontroller/1621407-presentedviewcontroller?language=objc the `presentedViewController` is only the top view controller when it is presented with `presentViewController:animated:completion:`. 
What if the top view controller is presented by [pushViewController](https://developer.apple.com/documentation/uikit/uinavigationcontroller/1621887-pushviewcontroller?language=objc)? Does this still work?

Comment (chris-rutkowski) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: @cyanglaz I am already repeating myself, but let me rephrase.

`presentViewController` is a method of `UIViewController`
UINavigationController, UITabBarController, MySweetViewController - they are all subclasses of UIViewController, therefore they can be presented on any other view controller, or some other view controller can be presented on top of them.

The top view controller **presented** by push view controller - such a thing doesn't exist.
You can have UINavigationController either as a root of the window, or presented on some other view controller. And Navigation Controller manages the stack of view controllers. These view controllers shouldn't present any other view controller (because they are managed by NC). Instead, you either push another VC in NC (horizontal), or present VC on top of NC (vertical).

Moreover - if UINavigationController is one of the children of UITabBarController, you shouldn't present any VC on top of that NC, you should present on top of UITabBarController instead.

Comment (cyanglaz) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: Ok. I'm interested in this case you presented 
```
example2:
window
  navigation controller (home, details)
     image picker
```

In this case, when you go through the code, what is the top view controller? Is it the `navigation controller`? 
And the image picker is going to use the navigation controller as the "viewController", regardless which view controller is the on the top of the stack of the navigation controller? Am I understanding correctly?

Comment (chris-rutkowski) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:44]: exactly, image picker will be correctly presented on navigation controller. The contents of navigation controller doesn't matter at all.

**Comment by chris-rutkowski** [comment]
@cyanglaz added unit test and improved the solution a bit.

BTW I cannot find any guide to run iOS tests. I had to create a dummy project which wasn't very easy.

**Comment by cyanglaz** [comment]
@chris-rutkowski 
try
```
cd ios
pod lib lint
```

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.h [hunk 3]
```diff
@@ -7,7 +7,7 @@
 @interface FLTImagePickerPlugin : NSObject <FlutterPlugin>
 
 // For testing only.
```
Comment (cyanglaz) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.h:9]: Should we remove this line now?

Comment (chris-rutkowski) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.h:9]: I don't think so? Both of these methods here are exposed only for testing and not meant to be used in normal circumstances.
